### ISSUE TITLE
Cancelled Orders With Successful Payments

### ISSUE THREAD

**Issue by sprankhub** [PAYONE-GmbH/magento-1#174]
We have the issue that orders are sometimes cancelled shortly after creation. A bit later, the payment is successfully made and the order is re-set to processing state. However, the product status is still cancelled, so that the order is in a state, where we cannot do anything with it: Since the products are marked as cancelled, Magento does not allow to create an invoice or a shipment. We have to manually un-cancel the products via a script or via the DB. The order status history looks like this:

![image](https://user-images.githubusercontent.com/930199/37084902-ff58c164-21f3-11e8-8c41-7ccf5d09f5fd.png)

We already contacted support and together, we came up with a process how the issue can be reproduced:

1. Add something to the cart, go to the checkout, choose a payment method (e.g. Sofort) and click on the order button.
2. You are redirected to the payment provider (e.g. Sofort). There, you click on "Abbrechen und zurück zum Händler". You are then redirected to the shopping cart and your order is cancelled (see https://github.com/PAYONE-GmbH/magento-1/blob/37eb52888d1317fe37db958dc8d492d09899898a/app/code/community/Payone/Core/controllers/Checkout/Onepage/PaymentController.php#L40-L50).
3. Use the browser back button multiple times until you are on the page of the payment provider again. Complete the payment.

Besides the fact that the order is in an inconsistent state, the user will see an error message even after the successful payment in step 3. He is redirected to the `payone_core/checkout_onepage_payment/success` URL. Unfortunately, instead of showing the order success page, the (filled!) cart is shown with the error message "Der Bestellvorgang wurde abgebrochen".

We use Magento 1.9.3.8 and Payone_Core 4.1.1.

**Comment by fjbender** [comment]
Thanks for the detailed bug report. We'll look into this.

**Comment by sprankhub** [comment]
When? Sorry for bugging you, but this is quite important IMHO.

**Comment by fjbender** [comment]
It's next on the list. These "back button" issues with redirect payment methods are always a pest, however, so I hope it won't take too much time to fix it.

**Comment by sprankhub** [comment]
Status? :-/

**Comment by FatchipRobert** [comment]
I checked the issue and I was able reproduce it.

Unfortunately there is not really an elegant solution for this.
At least I didnt find one yet.

I wasn't able to manipulate the redirect process so that it's not possible 
to jump back to the external payment provider with the browser-back-button.
This would have been the most elegant way but it didn't work.

The second best method would have been to automatically "uncancel" the order in that case.
But thats something the Magento core should provide and nothing that a 
payment module should implement individually with the given complexity that can be seen here:
https://magento.stackexchange.com/a/182442
Since the inventory needs to be reindexed when the stock is reduced again, 
this is nothing that should be done automatically during the checkout.

The only place where I could eventually see some kind of custom "uncancel" functionality 
would be a button in the backend for rare usage.

Since this is clearly user error I could also imagine showing warnings to not use the browser back button,
which could be shown before the redirect and/or on the cancel return page.

**Comment by sprankhub** [comment]
I agree that this issue is difficult to solve. But I strongly disagree that this is user error. The payment providers DO provide an option to "cancel and go back" (in the case of Sofort). Besides of that, why shouldn't I use the browser back button if I change my mind and do not want to pay via Sofort, but via PayPal instead? I think this is an absolutely valid user action. Hence, you / we have to support this functionality without any issues. If I see it correctly, the only valid option is that on cancel, the PAYONE module should:

1. cancel the order,
2. deactivate the quote / empty the cart,
3. add a message "hey, you just cancelled your order, but feel free to place a new one",
4. and redirect the user to the empty cart or the homepage.

Another nicer, but more complex solution:

1. On return, mark the quote as "not successfully paid".
2. If the user pays this quote later, create a new quote / order and do not re-use the old quote (which has been cancelled).

Maybe you can check how other payment providers deal with this issue...

**Comment by FatchipRobert** [comment]
The case with "the customer changes his mind and doesnt want to pay with Sofort, but with PayPal instead" is working though.
The first try with Sofort stays in the the order table as canceled and a new order for the PayPal payment is created.

Deactivating the quote and emptying the cart doesn't help at all unfortunately, because the customer can jump back into the "dead order process" and the shop can't do anything to prevent that, because the browser back button just jumps back in the browser-history and there is no shop-processing done in between.

Your second solution sounds like what the "correct way" of the customer changing his mind does - leaving the canceled order canceled and creating a new one.
I will look into that.

**Comment by sprankhub** [comment]
Status?

**Comment by fjbender** [comment]
We have a pull request #193 for this. However, it's still in testing and we don't know yet, whether it's working as expected. You are, as always, welcome to add your two cents. :)

**Comment by FatchipRobert** [comment]
The pull request is not 100% precise yet.
There should be a new one today or tomorrow.

**Comment by fjbender** [comment]
@sprankhub We now have updated #198. I wasn't quite able to force a re-login into Sofort's new interface (maybe they disabled it?). The PR is still in testing, your feedback is always welcome.

**Comment by sprankhub** [comment]
Do not get me wrong, but I fear the customer does not want to pay me for testing this in an unfinished state ;-) But I will definitely test it as soon as this is officially merged and released!

**Comment by fjbender** [comment]
Sure :) Could've been that you encountered a similar behavior with Sofort though. Will get back to you once we're done testing.

**Comment by fjbender** [comment]
We have now tested #198 and it fixes the problem, thus merged. We should be ready to close this issue.

**Comment by sprankhub** [comment]
Looking forward to the next release - I will test it then.

**Comment by phoenix-schwarz** [comment]
The changes from https://github.com/PAYONE-GmbH/magento-1/pull/198/commits/35435a09cec76dc6ea9a432a527a0213e04ffc7e#diff-893d7b14c3466eed0a5dda31e1635567 are not included in Version 4.1.5 and 4.1.6 - is this by purpose?

**Comment by fjbender** [comment]
bc1a1b5e693d2d8f2348669e3ee29c409ec2f53a overwrote these specific changes.

We're looking into that.

**Comment by fjbender** [comment]
This should be fixed with https://github.com/PAYONE-GmbH/magento-1/releases/tag/v4.1.7

@sprankhub Can you confirm?

**Comment by sprankhub** [comment]
I have a similar issue now with version 4.2.0, which looks related:

![image](https://user-images.githubusercontent.com/930199/45761670-f8969080-bc2c-11e8-9140-a5753fe9c03a.png)

It is basically the same: The order is cancelled, but the payment is confirmed afterwards. The only difference now is that the order stays in the canceled status. Any idea what that is?

**Comment by fjbender** [comment]
We didn't change anything here willingly. The only PR I could find remotely touching this subject is #260, but this shouldn't become active here.

However, I was unable to reproduce this behavior.

**Comment by HKSHendrikLegge** [comment]
I could observe that between the release of Version 4.2.0 and now the problems less often occurred. Only into the last two weeks the problems occurred more times again. may have been something changed by payone here which let the problems occurred less times in this time window?

**Comment by fjbender** [comment]
@HKSHendrikLegge Did you update to the current master? Or are you suspecting a platform change?

**Comment by phizab** [comment]
We've just updated to 4.5.0 and got the same problem with the first order. Sadly I cannot reproduce this and I hope, that there won't come more orders like this.

![payone_cancel](https://user-images.githubusercontent.com/1841607/74147139-f0186e80-4c02-11ea-85eb-9fb0fdbc7615.png)

Any news to this since the issue is still open?
